modules/java/generator/android-21/java/org/opencv/android/JavaCamera2View.java

```diff
@@ -0,0 +1,393 @@
+        }
+        return true;
+    }
+
+    private final CameraDevice.StateCallback mStateCallback = new CameraDevice.StateCallback() {
+
+        @Override
+        public void onOpened(CameraDevice cameraDevice) {
+            mCameraDevice = cameraDevice;
+            mCameraOpenCloseLock.release();
+            createCameraPreviewSession();
+        }
+
+        @Override
+        public void onDisconnected(CameraDevice cameraDevice) {
+            cameraDevice.close();
+            mCameraDevice = null;
+            mCameraOpenCloseLock.release();
+        }
+
+        @Override
+        public void onError(CameraDevice cameraDevice, int error) {
+            cameraDevice.close();
+            mCameraDevice = null;
+            mCameraOpenCloseLock.release();
+        }
+
+    };
+
+    private void createCameraPreviewSession() {
+        final int w=mPreviewSize.getWidth(), h=mPreviewSize.getHeight();
+        Log.i(LOGTAG, "createCameraPreviewSession("+w+"x"+h+")");
+        if(w<0 || h<0)
+            return;
+        try {
+            mCameraOpenCloseLock.acquire();
+            if (null == mCameraDevice) {
+                mCameraOpenCloseLock.release();
+                Log.e(LOGTAG, "createCameraPreviewSession: camera isn't opened");
+                return;
+            }
+            if (null != mCaptureSession) {
+                mCameraOpenCloseLock.release();
+                Log.e(LOGTAG, "createCameraPreviewSession: mCaptureSession is already started");
+                return;
+            }
+
+            mImageReader = ImageReader.newInstance(w,h,mPreviewFormat,2);
+            mImageReader.setOnImageAvailableListener(new ImageReader.OnImageAvailableListener() {
+                @Override
+                public void onImageAvailable(ImageReader reader) {
+                    Image image = reader.acquireLatestImage();
+                    if (image == null) return;
+
+                    // sanity checks - 3 planes
+                    Image.Plane[] planes = image.getPlanes();
+                    assert(planes.length == 3);
+                    assert(image.getFormat() == mPreviewFormat);
+
+                    // see also https://developer.android.com/reference/android/graphics/ImageFormat.html#YUV_420_888
+                    // Y plane (0) non-interleaved => stride == 1; U/V plane interleaved => stride == 2
+                    assert(planes[0].getPixelStride() == 1);
+                    assert(planes[1].getPixelStride() == 2);
+                    assert(planes[2].getPixelStride() == 2);
+
+                    ByteBuffer y_plane = planes[0].getBuffer();
+                    ByteBuffer uv_plane = planes[1].getBuffer();
+                    Mat y_mat = new Mat( h, w, CvType.CV_8UC1, y_plane );
+                    Mat uv_mat = new Mat( h/2, w, CvType.CV_8UC1, uv_plane );
+                    JavaCamera2Frame tempFrame = new JavaCamera2Frame(y_mat,uv_mat,w,h);
+                    deliverAndDrawFrame(tempFrame);
```

Comment:
Original `JavaCameraView` interface is asynchronous.
Image processing must be performed in a separate thread, see `CameraWorker`. This callback should just update the latest frame for next processing (non-processed frames are dropped).

Reply:
Yes, but the `onImageAvailable` listener will already be called within the `mBackgroundHandler` thread. Isn't this equivalent?

Reply:
Thread is "background" but it may block camera frames fetching and cause camera timeout/disconnection.
I have no experience with new Camera2 API, could you add some documentation link about this (callback processing/requirements)?

Reply:
See here: https://developer.android.com/reference/android/media/ImageReader.html#acquireLatestImage()
Since I'm calling `acquireLatestImage`, the queue will be emptied and only the most recent image will be passed on for processing. The documentation also mentions that "This function is recommended to use over acquireNextImage() for most use-cases, as it's more suited for real-time processing." 
To be on the safe side, I'll test what happens if the processing takes a really long time (let's say 500 ms).